src/Mod/Sketcher/Gui/ViewProviderSketch.h

```diff
@@ -38,6 +38,7 @@
 #include <QCoreApplication>
 #include <Gui/Document.h>
 #include "ShortcutListener.h"
+#include "TaskSketcherTool.h"

```

Comment:
Thanks for your feedbacks! 
For most points you raised I was expecting it. I did those because I didn't knew how it should be done properly.
For this it is because I didn't knew how to add my TaskSketcherTool class to the task bar from within sketchHandlers. Initially I wanted to open add the widget when a tool is launched and destroy it when it closes. But I couldn't find how to do it. So I did as the other widgets are added and declared the class in VP in order to be able to access it from the SketchHandler.

Reply:
I am still unsure about the tool in the taskbar. However, in order to do what you want, you do not need to modify the VP. You can access the task dialog from anywhere via:
```
Gui::TaskView::TaskDialog *dlg = Gui::Control().activeDialog();
TaskDlgEditSketch *sketchDlg = qobject_cast<TaskDlgEditSketch *>(dlg); 
```
So you can access the tool or some functionality of the tool via this pointer.

If the tools are going to be able to use this dialog, then it makes sense to give DrawSketchHandler the responsibility of registering/deregistering with the tool dialog. It can then provide a default implementation, which gets overriden by the specific DrawSketchHandler that wants to make use of the dialog.

By doing it this way, any tool gains the ability to use the tool widget and adapt its contents to the specific tool.

Sounds ok?

Reply:
Oh thanks I spent so much time looking for this... And I ended up doing this dirty hack.

You mean give it to the main DrawSketchHandler  class rather than each class? So I should declare this pointer in the main DrawSketchHandler  class, then every sub class will be able to access the pointer directly correct?

Reply:
I tried to do that and I'm trying to initialize dlg as follow in DrawSketchHandler constructor, but it's not working I'm getting an illegal storage access when a tool starts (on first attempt to reach for toolSettings).

Is that not correct?

`DrawSketchHandler::DrawSketchHandler() : sketchgui(0), dlg(Gui::Control().activeDialog()), toolSettings(qobject_cast<TaskSketcherTool*>(dlg)) {}`

Reply:
I think Gui::Control().activeDialog() may be the issue. Because the widget is not particularly active when you launch a tool. So why would it be considered to be the active dialog amongs all the Task widgets.

Reply:
Ok I misunderstood. So what I need is : 

`DrawSketchHandler::DrawSketchHandler() : sketchgui(0), toolSettings(qobject_cast<TaskDlgEditSketch*>(Gui::Control().activeDialog())->toolSettings) {}`
But this tells me that toolSettings is inacessible (because it's declared in protected.) So I can fix it in putting the declare in TaskDlgEditSketch.h public section. But I guess you're not going to like it :)